File: packages/native-helpers/audio-capture/Sources/AudioCapture/Debug/CaptureTraceWriter.swift
import AVFoundation
import CoreAudio
import Foundation

final class CaptureTraceWriter {
    private let directoryURL: URL
    private let jsonlURL: URL
    private var jsonHandle: FileHandle?
    private var channelHandles: [String: FileHandle] = [:]
    private var channelByteOffsets: [String: UInt64] = [:]
    private let lock = NSLock()
    private var isClosed = false

    init?(directoryPath: String) {
        self.directoryURL = URL(fileURLWithPath: directoryPath)
        self.jsonlURL = directoryURL.appendingPathComponent("native-capture-trace.jsonl")

        do {
            try FileManager.default.createDirectory(
                at: directoryURL,
                withIntermediateDirectories: true
            )
            FileManager.default.createFile(atPath: jsonlURL.path, contents: nil)
            self.jsonHandle = try FileHandle(forWritingTo: jsonlURL)
        } catch {
            Logger.error("Failed to initialize capture trace writer at \(jsonlURL.path): \(error.localizedDescription)")
            return nil
        }
    }

    deinit {
        try? close()
    }

    func record(event: String, metadata: [String: Any] = [:]) {
        lock.lock()
        defer { lock.unlock() }

        guard !isClosed, let jsonHandle else { return }

        var payload = metadata
        payload["event"] = event
        payload["loggedAtEpochMs"] = Int64((Date().timeIntervalSince1970 * 1000.0).rounded())

        do {
            let data = try JSONSerialization.data(withJSONObject: payload, options: [])
            try jsonHandle.write(contentsOf: data)
            try jsonHandle.write(contentsOf: Data([0x0A]))
        } catch {
            Logger.error("Failed to write capture trace event \(event): \(error.localizedDescription)")
        }
    }

    func recordSamples(
        event: String,
        channel: String,
        samples: [Float],
        metadata: [String: Any] = [:]
    ) {
        let payload = samples.withUnsafeBufferPointer { buffer in
            Data(buffer: buffer)
        }
        recordData(
            event: event,
            channel: channel,
            fileExtension: "f32le",
            payload: payload,
            sampleCount: samples.count,
            metadata: metadata
        )
    }

    func recordBytes(
        event: String,
        channel: String,
        payload: Data,
        metadata: [String: Any] = [:]
    ) {
        recordData(
            event: event,
            channel: channel,
            fileExtension: "bin",
            payload: payload,
            sampleCount: nil,
            metadata: metadata
        )
    }

    private func recordData(
        event: String,
        channel: String,
        fileExtension: String,
        payload: Data,
        sampleCount: Int?,
        metadata: [String: Any]
    ) {
        lock.lock()
        defer { lock.unlock() }

        guard !isClosed, let jsonHandle else { return }
        guard let appendResult = appendDataLocked(
            channel: channel,
            fileExtension: fileExtension,
            payload: payload
        ) else {
            return
        }

        var payload = metadata
        payload["event"] = event
        payload["loggedAtEpochMs"] = Int64((Date().timeIntervalSince1970 * 1000.0).rounded())
        payload["traceChannel"] = channel
        payload["traceFilePath"] = appendResult.filePath
        payload["traceByteOffset"] = Int64(appendResult.byteOffset)
        if let sampleCount {
            payload["traceSampleOffset"] = Int64(appendResult.byteOffset / 4)
            payload["traceSampleCount"] = sampleCount
        }

        do {
            let data = try JSONSerialization.data(withJSONObject: payload, options: [])
            try jsonHandle.write(contentsOf: data)
            try jsonHandle.write(contentsOf: Data([0x0A]))
        } catch {
            Logger.error("Failed to write capture trace sample event \(event): \(error.localizedDescription)")
        }
    }

    func close() throws {
        lock.lock()
        defer { lock.unlock() }

        guard !isClosed else { return }
        isClosed = true

        try jsonHandle?.synchronize()
        try jsonHandle?.close()
        jsonHandle = nil

        for (_, handle) in channelHandles {
            try handle.synchronize()
            try handle.close()
        }
        channelHandles.removeAll()
        channelByteOffsets.removeAll()
    }

    private func appendDataLocked(
        channel: String,
        fileExtension: String,
        payload: Data
    ) -> (filePath: String, byteOffset: UInt64)? {
        guard !payload.isEmpty else { return nil }

        let channelKey = "\(channel).\(fileExtension)"
        let fileURL = directoryURL.appendingPathComponent(channelKey)
        let handle: FileHandle
        if let existingHandle = channelHandles[channelKey] {
            handle = existingHandle
        } else {
            do {
                FileManager.default.createFile(atPath: fileURL.path, contents: nil)
                let newHandle = try FileHandle(forWritingTo: fileURL)
                channelHandles[channelKey] = newHandle
                channelByteOffsets[channelKey] = 0
                handle = newHandle
            } catch {
                Logger.error("Failed to initialize trace channel \(channel): \(error.localizedDescription)")
                return nil
            }
        }

        let byteOffset = channelByteOffsets[channelKey, default: 0]

        do {
            try handle.write(contentsOf: payload)
            channelByteOffsets[channelKey] = byteOffset + UInt64(payload.count)
            return (fileURL.path, byteOffset)
        } catch {
            Logger.error("Failed to append trace channel \(channel): \(error.localizedDescription)")
            return nil
        }
    }
}

func nonNilTraceFields(_ fields: [(String, Any?)]) -> [String: Any] {
    var result: [String: Any] = [:]
    for (key, value) in fields {
        if let value {
            result[key] = value
        }
    }
    return result
}

func captureSourceName(_ source: CaptureSource) -> String {
    switch source {
    case .micRaw:
        return "mic_raw"
    case .system:
        return "system"
    case .micProcessed:
        return "mic_processed"
    }
}

func audioTimeStampFields(
    _ timeStamp: UnsafePointer<AudioTimeStamp>?,
    prefix: String
) -> [String: Any] {
    guard let timeStamp else { return [:] }
    let value = timeStamp.pointee
    return [
        "\(prefix)SampleTime": value.mSampleTime,
        "\(prefix)HostTime": Int64(value.mHostTime),
        "\(prefix)RateScalar": value.mRateScalar,
        "\(prefix)WordClockTime": Int64(value.mWordClockTime),
        "\(prefix)Flags": Int(value.mFlags.rawValue),
        "\(prefix)SmpteSubframes": Int(value.mSMPTETime.mSubframes),
        "\(prefix)SmpteSubframeDivisor": Int(value.mSMPTETime.mSubframeDivisor),
        "\(prefix)SmpteCounter": Int(value.mSMPTETime.mCounter),
        "\(prefix)SmpteType": Int(value.mSMPTETime.mType.rawValue),
        "\(prefix)SmpteFlags": Int(value.mSMPTETime.mFlags.rawValue),
        "\(prefix)SmpteHours": Int(value.mSMPTETime.mHours),
        "\(prefix)SmpteMinutes": Int(value.mSMPTETime.mMinutes),
        "\(prefix)SmpteSeconds": Int(value.mSMPTETime.mSeconds),
        "\(prefix)SmpteFrames": Int(value.mSMPTETime.mFrames)
    ]
}

func avAudioFormatFields(
    _ format: AVAudioFormat?,
    prefix: String
) -> [String: Any] {
    guard let format else { return [:] }
    let streamDescription = format.streamDescription.pointee
    var fields: [String: Any] = [
        "\(prefix)SampleRate": format.sampleRate,
        "\(prefix)ChannelCount": Int(format.channelCount),
        "\(prefix)Interleaved": format.isInterleaved,
        "\(prefix)CommonFormat": commonFormat(from: streamDescription).map(commonFormatName) ?? "unknown",
        "\(prefix)FormatID": Int(streamDescription.mFormatID),
        "\(prefix)FormatFlags": Int(streamDescription.mFormatFlags),
        "\(prefix)BytesPerPacket": Int(streamDescription.mBytesPerPacket),
        "\(prefix)FramesPerPacket": Int(streamDescription.mFramesPerPacket),
        "\(prefix)BytesPerFrame": Int(streamDescription.mBytesPerFrame),
        "\(prefix)ChannelsPerFrame": Int(streamDescription.mChannelsPerFrame),
        "\(prefix)BitsPerChannel": Int(streamDescription.mBitsPerChannel)
    ]
    if let channelLayout = format.channelLayout {
        fields["\(prefix)ChannelLayoutTag"] = Int(channelLayout.layout.pointee.mChannelLayoutTag)
    }
    return fields
}

File: packages/native-helpers/audio-capture/windows/CaptureTraceWriter.cs
using System.Buffers;
using System.Runtime.InteropServices;
using System.Text.Json;

namespace AudioCapture.Windows;

internal sealed class CaptureTraceWriter : IDisposable
{
    private readonly string directoryPath;
    private readonly FileStream jsonStream;
    private readonly object traceLock = new();
    private readonly Dictionary<string, FileStream> channelStreams = new();
    private readonly Dictionary<string, long> channelByteOffsets = new();
    private bool closed;

    public CaptureTraceWriter(string directoryPath)
    {
        this.directoryPath = directoryPath;
        Directory.CreateDirectory(directoryPath);
        jsonStream = new FileStream(
            Path.Combine(directoryPath, "native-capture-trace.jsonl"),
            FileMode.Create,
            FileAccess.Write,
            FileShare.Read
        );
    }

    public void Record(string eventName, IReadOnlyDictionary<string, object?>? metadata = null)
    {
        lock (traceLock)
        {
            if (closed)
            {
                return;
            }

            var payload = CopyMetadata(metadata);
            payload["event"] = eventName;
            payload["loggedAtEpochMs"] = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            WriteJsonLineLocked(payload);
        }
    }

    public void RecordSamples(
        string eventName,
        string channel,
        ReadOnlySpan<float> samples,
        IReadOnlyDictionary<string, object?>? metadata = null
    )
    {
        if (samples.IsEmpty)
        {
            return;
        }

        var byteCount = samples.Length * sizeof(float);
        var rented = ArrayPool<byte>.Shared.Rent(byteCount);
        try
        {
            var payload = rented.AsSpan(0, byteCount);
            MemoryMarshalShim.CopyFloat32LittleEndian(samples, payload);
            RecordData(eventName, channel, "f32le", payload, samples.Length, metadata);
        }
        finally
        {
            ArrayPool<byte>.Shared.Return(rented);
        }
    }

    public void RecordBytes(
        string eventName,
        string channel,
        ReadOnlySpan<byte> payload,
        IReadOnlyDictionary<string, object?>? metadata = null
    )
    {
        RecordData(eventName, channel, "bin", payload, null, metadata);
    }

    public void Dispose()
    {
        lock (traceLock)
        {
            if (closed)
            {
                return;
            }

            closed = true;
            jsonStream.Flush();
            jsonStream.Dispose();
            foreach (var stream in channelStreams.Values)
            {
                stream.Flush();
                stream.Dispose();
            }

            channelStreams.Clear();
            channelByteOffsets.Clear();
        }
    }

    private void RecordData(
        string eventName,
        string channel,
        string fileExtension,
        ReadOnlySpan<byte> payloadBytes,
        int? sampleCount,
        IReadOnlyDictionary<string, object?>? metadata
    )
    {
        if (payloadBytes.IsEmpty)
        {
            return;
        }

        lock (traceLock)
        {
            if (closed)
            {
                return;
            }

            var appendResult = AppendDataLocked(channel, fileExtension, payloadBytes);
            var payload = CopyMetadata(metadata);
            payload["event"] = eventName;
            payload["loggedAtEpochMs"] = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            payload["traceChannel"] = channel;
            payload["traceFilePath"] = appendResult.FilePath;
            payload["traceByteOffset"] = appendResult.ByteOffset;

            if (sampleCount != null)
            {
                payload["traceSampleOffset"] = appendResult.ByteOffset / sizeof(float);
                payload["traceSampleCount"] = sampleCount.Value;
            }

            WriteJsonLineLocked(payload);
        }
    }

    private (string FilePath, long ByteOffset) AppendDataLocked(
        string channel,
        string fileExtension,
        ReadOnlySpan<byte> payload
    )
    {
        var channelKey = $"{channel}.{fileExtension}";
        var filePath = Path.Combine(directoryPath, channelKey);
        if (!channelStreams.TryGetValue(channelKey, out var stream))
        {
            stream = new FileStream(filePath, FileMode.Create, FileAccess.Write, FileShare.Read);
            channelStreams[channelKey] = stream;
            channelByteOffsets[channelKey] = 0;
        }

        var byteOffset = channelByteOffsets[channelKey];
        stream.Write(payload);
        channelByteOffsets[channelKey] = byteOffset + payload.Length;
        return (filePath, byteOffset);
    }

    private void WriteJsonLineLocked(IReadOnlyDictionary<string, object?> payload)
    {
        JsonSerializer.Serialize(jsonStream, payload);
        jsonStream.WriteByte(0x0A);
    }

    private static Dictionary<string, object?> CopyMetadata(
        IReadOnlyDictionary<string, object?>? metadata
    )
    {
        var payload = new Dictionary<string, object?>();
        if (metadata == null)
        {
            return payload;
        }

        foreach (var (key, value) in metadata)
        {
            if (value != null)
            {
                payload[key] = value;
            }
        }

        return payload;
    }
}

internal static class MemoryMarshalShim
{
    public static void CopyFloat32LittleEndian(ReadOnlySpan<float> source, Span<byte> destination)
    {
        var bytes = MemoryMarshal.AsBytes(source);
        bytes.CopyTo(destination);
    }
}
